[R2] presigned URLs learning page (#6449)
cloudflare/cloudflare-docs: content/r2/data-access/s3-api/presigned-urls.md

---
title: Presigned URLs
pcx_content_type: concept
---

# Presigned URLs

> A presigned URL authorizes anyone with the URL to perform an action to the S3 endpoint for an R2 bucket. By default, the S3 endpoint requires an `AUTHORIZATION` header signed by your token. Each presigned URL contains S3 parameters and search parameters containing the signature information that would be present in an `AUTHORIZATION` header. The performable action is restricted to a specific resource, an [operation](/r2/data-access/s3-api/api/), and has an associated timeout.

Reviewer: > S3 parameters and search parameters

This is a bit wordy. Just "Each presigned URL contains S3 parameters and search parameters containing the signature information that would be present in an `AUTHORIZATION` header"?

The S3 parameters that would be in the URL would be there either way and are unchanged.

> A presigned URL authorizes anyone with the URL to perform an action to the S3 endpoint for an R2 bucket. By default, the S3 endpoint requires an `AUTHORIZATION` header signed by your token. Each presigned URL contains S3 parameters and search parameters containing the signature information that would be present in an `AUTHORIZATION` header. The performable action is restricted to a specific resource, an [operation](/r2/data-access/s3-api/api/), and has an associated timeout.

Reviewer: "S3 API endpoint" / "S3 compatibility endpoint" / "S3 API compatibility endpoint" instead of "S3 endpoint"?

I would also link that to https://developers.cloudflare.com/r2/data-access/s3-api/

> A presigned URL authorizes anyone with the URL to perform an action to the S3 endpoint for an R2 bucket. By default, the S3 endpoint requires an `AUTHORIZATION` header signed by your token. Each presigned URL contains S3 parameters and search parameters containing the signature information that would be present in an `AUTHORIZATION` header. The performable action is restricted to a specific resource, an [operation](/r2/data-access/s3-api/api/), and has an associated timeout.

Author: ```suggestion
Presigned URLs are an [S3 concept](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html) for sharing direct access to your bucket without revealing your token secret. A presigned URL authorizes anyone with the URL to perform an action to the S3 endpoint for an R2 bucket. By default, the S3 endpoint requires an `AUTHORIZATION` header signed by your token. Each presigned URL contains S3 parameters and search parameters containing the signature information that would be present in an `AUTHORIZATION` header. The performable action is restricted to a specific resource, an [operation](/r2/data-access/s3-api/api/), and has an associated timeout.
```

There are three kinds of resources in R2: 

> 1. **Account**: For account-level operations (such as `CreateBucket`, `ListBuckets`, `DeleteBucket`) the identifier is the account ID.

Reviewer: You know reading this again, we don't group our API on the referenced page in this way. I wouldn't necessarily block this review to update that, but we probably want to clean up https://developers.cloudflare.com/r2/data-access/s3-api/api/ to match this (right now `CreateBucket` is in "bucket operations" and `ListObjects` is in "object operations" when they're not actually).

> 2. **Bucket**: For bucket-level operations (such as `ListObjects`, `PutBucketCors`) the identifier is the account ID, bucket name.

Reviewer: Do you think it's clear that it's ID and bucket name together identify it? I guess below we do say "all parts of the identifier" so probably ok....

3. **Object**: For object-level operations (such as `GetObject`, `PutObject`, `CreateMultipartUpload`) the identifier is the account ID, bucket name, object path.

> All parts of the identifier are part of the presigned URL. You cannot change the resource being accessed after the request is signed. For example, trying to change the bucket name to access the same object in a different bucket will not work.

Reviewer: "will not work" -> "will return a `403` with an error code of `SignatureDoesNotMatch`" as @KianNH mentioned.

> All parts of the identifier are part of the presigned URL. You cannot change the resource being accessed after the request is signed. For example, trying to change the bucket name to access the same object in a different bucket will not work.

Author: ```suggestion
All parts of the identifier are part of the presigned URL. You cannot change the resource being accessed after the request is signed. For example, trying to change the bucket name to access the same object in a different bucket will return a `403` with an error code of `SignatureDoesNotMatch`.
```

> Presigned URLs must have a defined expiry date. You can set an expiry date for your presigned URL from one second to 7 days in the future.  

Reviewer: In typical SDKs you provide a timeout and it signs the current date + timeout in the URL. Might be worthwhile to clarify it here with something like:

> Presigned URLs must have a defined expiry. You set a timeout in seconds up to 7 days (604800 seconds) into the future. The URL will contain the time when the URL was generated (`X-Amz-Date`) and the timeout (`X-Amz-Expires`) as search parameters. These search parameters are signed so tampering with them will result in `403` with an error code of `SignatureDoesNotMatch`.

> Presigned URLs must have a defined expiry date. You can set an expiry date for your presigned URL from one second to 7 days in the future.  

Author: ```suggestion
Presigned URLs must have a defined expiry. You can set a timeout from one second to 7 days (604,800 seconds) into the future. The URL will contain the time when the URL was generated (`X-Amz-Date`) and the timeout (`X-Amz-Expires`) as search parameters. These search parameters are signed so tampering with them will result in `403` with an error code of `SignatureDoesNotMatch`.
```

> Presigned URLs are generated offline with no communication with R2 and must be generated by an application with access to your R2 bucket's credentials.

Reviewer: I think we can omit the offline bit:

> Presigned URLs are generated with no communication with R2

e.g. if your signing application is a Worker, it will likely be confusing to a novice reader what "offline" means there.

> Presigned URLs are generated offline with no communication with R2 and must be generated by an application with access to your R2 bucket's credentials.

Author: ```suggestion
Presigned URLs are generated with no communication with R2 and must be generated by an application with access to your R2 bucket's credentials.
```

A valid alternative design to presigned URLs is to use a Worker with a [binding](/workers/platform/bindings/) that implements your security policy.

{{<Aside type="note" header="Bindings">}}

A binding is a how your Worker interacts with external resources such as [KV Namespaces](/workers/runtime-apis/kv/), [Durable Objects](/workers/runtime-apis/durable-objects/), or [R2 Buckets](#api). A binding is a runtime variable that the Workers runtime provides to your code. You can declare a variable name in your `wrangler.toml` file that will be bound to these resources at runtime, and interact with them through this variable. Every binding's variable name and behavior is determined by you when deploying the Worker. Refer to [Environment Variables](/workers/platform/environment-variables) for more information.

A binding is defined in the `wrangler.toml` file of your Worker project's directory.

{{</Aside>}}

A possible use case may be restricting an application to only be able to upload to a specific URL. With presigned URLs, your central signing application might look like the following JavaScript code running on Cloudflare Workers, workerd, or another platform:

> ```

Reviewer 2: ```suggestion
```ts
```

In some cases, Workers lets you implement certain functionality more easily. For example, if you wanted to offer a write-once guarantee so that users can only upload to a path once, with pre-signed URLs, you would need to sign specific headers and require the sender to send them. This adds some complexity:

```
    const signedUrl = await r2.sign(new Request(
      `https://<DROPBOX BUCKET><ACCOUNT ID>.r2.cloudflarestorage.com${url.pathname}`, {
        method: 'PUT'
      }
    ), {
      signQuery: true,
      headers: {
        // The signed request is valid for 1 hour.
        'X-Amz-Expires': 3600,
        'If-Unmodified-Since': 'Tue, 28 Sep 2021 16:00:00 GMT',
      },
    });
```

Note that the caller has to add the the same `If-Unmodified-Since` header to use the URL. The caller cannot omit the header or use a different header.

In a Worker, you would change your upload to:

> ```

Reviewer 2: ```suggestion
```ts
```

All parts of the identifier are part of the presigned URL. You cannot change the resource being accessed after the request is signed. For example, trying to change the bucket name to access the same object in a different bucket will return a `403` with an error code of `SignatureDoesNotMatch`.

Presigned URLs must have a defined expiry. You can set a timeout from one second to 7 days (604,800 seconds) into the future. The URL will contain the time when the URL was generated (`X-Amz-Date`) and the timeout (`X-Amz-Expires`) as search parameters. These search parameters are signed so tampering with them will result in `403` with an error code of `SignatureDoesNotMatch`.

Presigned URLs are generated with no communication with R2 and must be generated by an application with access to your R2 bucket's credentials.

## Presigned URL use cases

There are three ways to grant an application access to R2:

1. The application has its own copy of an [R2 API token](/r2/data-access/s3-api/tokens/).
2. The application requests a copy of an R2 API token from a vault application and promises to not permanently store that token locally.

> 3. The application requests a central application to give it a presigned URL so it can use to perform an action.

Reviewer 3: Not trying to nit, but the "so" threw me off on the first couple reads. I think you can remove it.

      `https://<DROPBOX BUCKET>.<ACCOUNT ID>.r2.cloudflarestorage.com${url.pathname}`, {
        method: 'PUT'
      }
    ), {
      aws: { signQuery: true },
      headers: {
        // The signed request is valid for 1 hour.
        'X-Amz-Expires': 3600,
      },
    });

    // Caller can now use this URL to upload to that object.
    return new Response(signedUrl, { status: 200 });
  }

  ... handle other kinds of requests
}
```

An equivalent Cloudflare Worker would be:

```ts
interface Env {
  DROP_BOX_BUCKET: R2Bucket
}

export default <ExportedHandler<Env>> {
  async fetch(req, env) {
  const authorization = await authorize(request);
  if (authorization === undefined) {
    return new Response('invalid authorization', { status: 401 })
  }
  if (authorization.application === 'drop-box') {
    const url = new URL(request.url);

    // Check that the application is only requesting a URL
    if (url.pathname.startsWith('/my-upload-folder/')) {
      return new Response('invalid destination folder', { status: 403 });
    }

    await env.DROP_BOX_BUCKET.put(url.toString().substring(1), request.body);

    return new Response(null, { status: 200 });
  }

> //  ... handle other kinds of requests

Author: ```suggestion
  //  ... handle other kinds of requests
```

* Enterprise customers can upload 500 MiB by default and can ask their account team to raise this limit.
* Detecting [precondition failures](/r2/data-access/s3-api/extensions/#conditional-operations-in-putobject) is currently easier with presigned URLs as compared with R2 bindings.

Note that these limitations depends on R2's extension for conditional uploads. Amazon's S3 service does not offer such functionality at this time.

## Presigned URLs versus Public Buckets

Presigned URLs share some superificial similarity with public buckets. If you give out presigned URLs only for `GET`/`HEAD` operations on specific objects in a bucket, then your presigned URL functionality is mostly similar to public buckets. The notable exception is that any custom metadata associated with the object is rendered in headers with the `x-amz-meta-` prefix. Any error responses are returned as XML documents, as they would with normal non-presigned S3 access. Presigned URLs can be generated for any S3 operation. After a presigned URL is generated it can be reused as many times as the holder of the URL wants until the signed expiry date.

> [Public buckets](/r2/data-access/public-buckets/) are available on a regular HTTP endpoint. By default, there is no authorization or access controls associated with a public bucket. Anyone with a public bucket URL can access an object in that public bucket. If you are using a custom domain to expose the R2 bucket, you can manage authorization and access controls as you would for a Cloudflare zone. Public buckets only provide `GET`/`HEAD` on a known object path. Public bucket errors are rendered as HTML pages. There is no ability to do S3 operations such as `CreateBucket`, `ListBuckets`, or `DeleteBucket`.

Reviewer 3: "There is no ability to perform S3 options such as `CreateBucket`,...." ?

---
title: Presigned URLs
pcx_content_type: concept
---

# Presigned URLs

Presigned URLs are an [S3 concept](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html) for sharing direct access to your bucket without revealing your token secret. A presigned URL authorizes anyone with the URL to perform an action to the S3 compatibility endpoint for an R2 bucket. By default, the S3 endpoint requires an `AUTHORIZATION` header signed by your token. Every presigned URL has S3 parameters and search parameters containing the signature information that would be present in an `AUTHORIZATION` header. The performable action is restricted to a specific resource, an [operation](/r2/data-access/s3-api/api/), and has an associated timeout.

There are three kinds of resources in R2: 

1. **Account**: For account-level operations (such as `CreateBucket`, `ListBuckets`, `DeleteBucket`) the identifier is the account ID.
2. **Bucket**: For bucket-level operations (such as `ListObjects`, `PutBucketCors`) the identifier is the account ID, bucket name.

> 3. **Object**: For object-level operations (such as `GetObject`, `PutObject`, `CreateMultipartUpload`) the identifier is the account ID, bucket name, object path.

Author: ```suggestion
3. **Object**: For object-level operations (such as `GetObject`, `PutObject`, `CreateMultipartUpload`) the identifier is the account ID, bucket name, and object path.
```

All parts of the identifier are part of the presigned URL. You cannot change the resource being accessed after the request is signed. For example, trying to change the bucket name to access the same object in a different bucket will return a `403` with an error code of `SignatureDoesNotMatch`.

Presigned URLs must have a defined expiry. You can set a timeout from one second to 7 days (604,800 seconds) into the future. The URL will contain the time when the URL was generated (`X-Amz-Date`) and the timeout (`X-Amz-Expires`) as search parameters. These search parameters are signed so tampering with them will result in `403` with an error code of `SignatureDoesNotMatch`.

Presigned URLs are generated with no communication with R2 and must be generated by an application with access to your R2 bucket's credentials.

## Presigned URL use cases

There are three ways to grant an application access to R2:

1. The application has its own copy of an [R2 API token](/r2/data-access/s3-api/tokens/).
2. The application requests a copy of an R2 API token from a vault application and promises to not permanently store that token locally.

> 3. The application requests a central application to give it a presigned URL so it can use to perform an action.

Author: ```suggestion
3. The application requests a central application to give it a presigned URL it can use to perform an action.
```

- [AWS SDK for JS](/r2/examples/aws-sdk-js/#generate-presigned-urls)
- [AWS SDK for PHP](/r2/examples/aws-sdk-php/#generate-presigned-urls)
- [AWS CLI](/r2/examples/aws-cli/#generate-presigned-urls)

## Presigned URL alternative with Workers

A valid alternative design to presigned URLs is to use a Worker with a [binding](/workers/platform/bindings/) that implements your security policy.

{{<Aside type="note" header="Bindings">}}

A binding is a how your Worker interacts with external resources such as [KV Namespaces](/workers/runtime-apis/kv/), [Durable Objects](/workers/runtime-apis/durable-objects/), or [R2 Buckets](#api). A binding is a runtime variable that the Workers runtime provides to your code. You can declare a variable name in your `wrangler.toml` file that will be bound to these resources at runtime, and interact with them through this variable. Every binding's variable name and behavior is determined by you when deploying the Worker. Refer to [Environment Variables](/workers/platform/environment-variables) for more information.

A binding is defined in the `wrangler.toml` file of your Worker project's directory.

{{</Aside>}}

> A possible use case may be restricting an application to only be able to upload to a specific URL. With presigned
> URLs, your central signing application might look like the following JavaScript code running on Cloudflare Workers, workerd,
> or another platform:

Author: ```suggestion
A possible use case may be restricting an application to only be able to upload to a specific URL. With presigned URLs, your central signing application might look like the following JavaScript code running on Cloudflare Workers, workerd, or another platform:
```

import { AwsClient } from 'aws4fetch'

const r2 = new AwsClient({ accessKeyId: '', secretAccessKey: '', service: 's3' , region: 'auto' });

export async function fetch(request: Request): Promise<Response> {

>   const authorization = await authorize(request);

Reviewer 2: Note: `authorize` is not provided in the Worker and is just an indication to the user that they should do authorisation (i.e a pre shared secret).

export async function fetch(request: Request): Promise<Response> {
  const authorization = await authorize(request);
  if (authorization === undefined) {
    return new Response('invalid authorization', { status: 401 })
  }
  if (authorization.application === 'drop-box') {
    const url = new URL(request.url);

    // Check that the application is only requesting a URL
    if (url.pathname.startsWith('/my-upload-folder/')) {
      return new Response('invalid destination folder', { status: 403 });
    }

    const signedUrl = await r2.sign(new Request(
      `https://<DROPBOX BUCKET><ACCOUNT ID>.r2.cloudflarestorage.com${url.pathname}`, {
        method: 'PUT'
      }
    ), {
      signQuery: true,
      headers: {
        // The signed request is valid for 1 hour.
        'X-Amz-Expires': 3600,
      },
    });

    // Caller can now use this URL to upload to that object.
    return new Response(signedUrl, { status: 200 });
  }

  ... handle other kinds of requests
}
```

An equivalent Cloudflare Worker would be:

```
interface Env {
  DROP_BOX_BUCKET: R2Bucket
}

> export async function fetch(request: Request, env: Env): Promise<Response> {

Reviewer 2: My `<Environment>` would be `<Env>` to match the interface above - just typed Environment on autopilot.

> import { AwsClient } from 'aws4fetch'
> 
> const r2 = new AwsClient({ accessKeyId: '', secretAccessKey: '', service: 's3' , region: 'auto' });
> 
> export async function fetch(request: Request): Promise<Response> {
>   const authorization = await authorize(request);
>   if (authorization === undefined) {
>     return new Response('invalid authorization', { status: 401 })
>   }
>   if (authorization.application === 'drop-box') {
>     const url = new URL(request.url);
> 
>     // Check that the application is only requesting a URL
>     if (url.pathname.startsWith('/my-upload-folder/')) {
>       return new Response('invalid destination folder', { status: 403 });
>     }
> 
>     const signedUrl = await r2.sign(new Request(
>       `https://<DROPBOX BUCKET>.<ACCOUNT ID>.r2.cloudflarestorage.com${url.pathname}`, {
>         method: 'PUT'
>       }
>     ), {
>       aws: { signQuery: true },
>       headers: {
>         // The signed request is valid for 1 hour.
>         'X-Amz-Expires': 3600,
>       },
>     });
> 
>     // Caller can now use this URL to upload to that object.
>     return new Response(signedUrl, { status: 200 });
>   }
> 
>   ... handle other kinds of requests
> }

Reviewer 2: ```suggestion
import { AwsClient } from 'aws4fetch';

const r2 = new AwsClient({
	accessKeyId: '',
	secretAccessKey: '',
	service: 's3',
	region: 'auto',
});

export default <ExportedHandler>{
	async fetch(request) {
		const authorization = await authorize(request);
		if (authorization === undefined) {
			return new Response('invalid authorization', { status: 401 });
		}
		if (authorization.application === 'drop-box') {
			const url = new URL(request.url);

			// Check that the application is only requesting a URL
			if (url.pathname.startsWith('/my-upload-folder/')) {
				return new Response('invalid destination folder', { status: 403 });
			}

			const signedUrl = await r2.sign(
				new Request(
					`https://<DROPBOX BUCKET>.<ACCOUNT ID>.r2.cloudflarestorage.com${url.pathname}`,
					{
						method: 'PUT',
					}
				),
				{
					aws: { signQuery: true },
					headers: {
						// The signed request is valid for 1 hour.
						'X-Amz-Expires': 3600,
					},
				}
			);

			// Caller can now use this URL to upload to that object.
			return new Response(signedUrl, { status: 200 });
		}

		// ... handle other kinds of requests
	},
};
```

> interface Env {
>   DROP_BOX_BUCKET: R2Bucket
> }
> 
> export default <ExportedHandler<Env>> {
>   async fetch(req, env) {
>   const authorization = await authorize(request);
>   if (authorization === undefined) {
>     return new Response('invalid authorization', { status: 401 })
>   }
>   if (authorization.application === 'drop-box') {
>     const url = new URL(request.url);
> 
>     // Check that the application is only requesting a URL
>     if (url.pathname.startsWith('/my-upload-folder/')) {
>       return new Response('invalid destination folder', { status: 403 });
>     }
> 
>     await env.DROP_BOX_BUCKET.put(url.toString().substring(1), request.body);
> 
>     return new Response(null, { status: 200 });
>   }
> 
> //  ... handle other kinds of requests
>   }
> }

Reviewer 2: ```suggestion
interface Env {
	DROP_BOX_BUCKET: R2Bucket;
}

export default <ExportedHandler<Env>>{
	async fetch(req, env) {
		const authorization = await authorize(request);
		if (authorization === undefined) {
			return new Response('invalid authorization', { status: 401 });
		}
		if (authorization.application === 'drop-box') {
			const url = new URL(request.url);

			// Check that the application is only requesting a URL
			if (url.pathname.startsWith('/my-upload-folder/')) {
				return new Response('invalid destination folder', { status: 403 });
			}

			await env.DROP_BOX_BUCKET.put(url.toString().substring(1), request.body);

			return new Response(null, { status: 200 });
		}

		//  ... handle other kinds of requests
	},
};
```

```

Notice the total absence of any configuration or token secrets present in the Worker code. Instead, you would create a `wrangler.toml` [binding](/r2/data-access/workers-api/workers-api-usage/#4-bind-your-bucket-to-a-worker) to whatever bucket represents the bucket you will upload to. Additionally, authorization is handled in-line with the upload which can reduce latency.

In some cases, Workers lets you implement certain functionality more easily. For example, if you wanted to offer a write-once guarantee so that users can only upload to a path once, with pre-signed URLs, you would need to sign specific headers and require the sender to send them. This adds some complexity:

```ts

>     const signedUrl = await r2.sign(new Request(
>       `https://<DROPBOX BUCKET>.<ACCOUNT ID>.r2.cloudflarestorage.com${url.pathname}`, {
>         method: 'PUT'
>       }
>     ), {
>       aws: { signQuery: true },
>       headers: {
>         // The signed request is valid for 1 hour.
>         'X-Amz-Expires': 3600,
>         'If-Unmodified-Since': 'Tue, 28 Sep 2021 16:00:00 GMT',
>       },
>     });

Reviewer 2: ```suggestion
const signedUrl = await r2.sign(
	new Request(
		`https://<DROPBOX BUCKET>.<ACCOUNT ID>.r2.cloudflarestorage.com${url.pathname}`,
		{
			method: 'PUT',
		}
	),
	{
		aws: { signQuery: true },
		headers: {
			// The signed request is valid for 1 hour.
			'X-Amz-Expires': 3600,
			'If-Unmodified-Since': 'Tue, 28 Sep 2021 16:00:00 GMT',
		},
	}
);
```

```

Note that the caller has to add the the same `If-Unmodified-Since` header to use the URL. The caller cannot omit the header or use a different header.

In a Worker, you would change your upload to:

```
    const existingObject = await env.DROP_BOX_BUCKET.put(url.substring(1), request.body, {
      onlyIf: {
        // No objects will have been uploaded before September 28th, 2021 which
        // is the initial R2 announcement.
        uploadedBefore: new Date(1632844800000),
      }
    });
    if (existingObject?.etag !== request.headers.get('etag')) {
      return new Response('attempt to overwrite object', { status: 400 })
    }
```

Cloudflare Workers currently have some limitations that you may need to consider:

* You cannot upload more than 100 MiB (200 MiB for Business customers) to a Worker.
* Enterprise customers can upload 500 MiB by default and can ask their account team to raise this limit.

> * Detecting the precondition failure is currently easier with presigned URLs as compared with R2 bindings.

Author: ```suggestion
* Detecting [precondition failures](/r2/data-access/s3-api/extensions/#conditional-operations-in-putobject) is currently easier with presigned URLs as compared with R2 bindings.
```

>     const existingObject = await env.DROP_BOX_BUCKET.put(url.toString().substring(1), request.body, {
>       onlyIf: {
>         // No objects will have been uploaded before September 28th, 2021 which
>         // is the initial R2 announcement.
>         uploadedBefore: new Date(1632844800000),
>       }
>     });
>     if (existingObject?.etag !== request.headers.get('etag')) {
>       return new Response('attempt to overwrite object', { status: 400 })
>     }

Reviewer 2: ```suggestion
const existingObject = await env.DROP_BOX_BUCKET.put(
	url.toString().substring(1),
	request.body,
	{
		onlyIf: {
			// No objects will have been uploaded before September 28th, 2021 which
			// is the initial R2 announcement.
			uploadedBefore: new Date(1632844800000),
		},
	}
);
if (existingObject?.etag !== request.headers.get('etag')) {
	return new Response('attempt to overwrite object', { status: 400 });
}
```

Note that these limitations depends on R2's extension for conditional uploads. Amazon's S3 service does not offer such functionality at this time.

## Presigned URLs versus Public Buckets

> Presigned URLs share some superificial similarity with public buckets. If you hand out presigned URLs only for `GET`/`HEAD` operations on specific objects of a bucket, then your presigned URL functionality is mostly similar to public buckets. The notable exception is that any custom metadata associated with the object is rendered in headers with the `x-amz-meta-` prefix. Any error responses are returned as XML documents, as they would with normal non-presigned S3 access. Presigned URLs can be generated for any S3 operation. After a presigned URL is generated it can be reused as many times as the holder of the URL wants until the signed expiry date.

Author: ```suggestion
Presigned URLs share some superificial similarity with public buckets. If you give out presigned URLs only for `GET`/`HEAD` operations on specific objects in a bucket, then your presigned URL functionality is mostly similar to public buckets. The notable exception is that any custom metadata associated with the object is rendered in headers with the `x-amz-meta-` prefix. Any error responses are returned as XML documents, as they would with normal non-presigned S3 access. Presigned URLs can be generated for any S3 operation. After a presigned URL is generated it can be reused as many times as the holder of the URL wants until the signed expiry date.
```

[Public buckets](/r2/data-access/public-buckets/) are available on a regular HTTP endpoint. By default, there is no authorization or access controls associated with a public bucket. Anyone with a public bucket URL can access an object in that public bucket. If you are using a custom domain to expose the R2 bucket, you can manage authorization and access controls as you would for a Cloudflare zone. Public buckets only provide `GET`/`HEAD` on a known object path. Public bucket errors are rendered as HTML pages. There is no ability to do S3 operations such as `CreateBucket`, `ListBuckets`, or `DeleteBucket`.

> Choosing between presigned URLs and public buckets is dependent on your specific use-case. You can also use both if your architecture should use public buckets in one situation and presigned URLs in another. It is useful to note that presigned URLs will expose your account ID and bucket name to whoever gets a copy of the URL. Public bucket URLs do not contain the account ID or bucket name. Typically, you will not share presigned URLs directly with end users or browsers, as presigned URLs are used more for internal applications.

Author: ```suggestion
Choosing between presigned URLs and public buckets is dependent on your specific use case. You can also use both if your architecture should use public buckets in one situation and presigned URLs in another. It is useful to note that presigned URLs will expose your account ID and bucket name to whoever gets a copy of the URL. Public bucket URLs do not contain the account ID or bucket name. Typically, you will not share presigned URLs directly with end users or browsers, as presigned URLs are used more for internal applications.
```

A binding is a how your Worker interacts with external resources such as [KV Namespaces](/workers/runtime-apis/kv/), [Durable Objects](/workers/runtime-apis/durable-objects/), or [R2 Buckets](#api). A binding is a runtime variable that the Workers runtime provides to your code. You can declare a variable name in your `wrangler.toml` file that will be bound to these resources at runtime, and interact with them through this variable. Every binding's variable name and behavior is determined by you when deploying the Worker. Refer to [Environment Variables](/workers/platform/environment-variables) for more information.

A binding is defined in the `wrangler.toml` file of your Worker project's directory.

{{</Aside>}}

A possible use case may be restricting an application to only be able to upload to a specific URL. With presigned URLs, your central signing application might look like the following JavaScript code running on Cloudflare Workers, workerd, or another platform:

```ts
import { AwsClient } from 'aws4fetch';

const r2 = new AwsClient({
	accessKeyId: '',
	secretAccessKey: '',
	service: 's3',
	region: 'auto',
});

export default <ExportedHandler>{
	async fetch(request) {

> 		const authorization = await authorize(request);

Author: ```suggestion
		const authorization = await authorize(request);
// `authorize` is not provided in the Worker. It is an indication that you should do authorization (that is, a preshared secret)
```

* Enterprise customers can upload 500 MiB by default and can ask their account team to raise this limit.
* Detecting [precondition failures](/r2/data-access/s3-api/extensions/#conditional-operations-in-putobject) is currently easier with presigned URLs as compared with R2 bindings.

Note that these limitations depends on R2's extension for conditional uploads. Amazon's S3 service does not offer such functionality at this time.

## Presigned URLs versus Public Buckets

Presigned URLs share some superificial similarity with public buckets. If you give out presigned URLs only for `GET`/`HEAD` operations on specific objects in a bucket, then your presigned URL functionality is mostly similar to public buckets. The notable exception is that any custom metadata associated with the object is rendered in headers with the `x-amz-meta-` prefix. Any error responses are returned as XML documents, as they would with normal non-presigned S3 access. Presigned URLs can be generated for any S3 operation. After a presigned URL is generated it can be reused as many times as the holder of the URL wants until the signed expiry date.

> [Public buckets](/r2/data-access/public-buckets/) are available on a regular HTTP endpoint. By default, there is no authorization or access controls associated with a public bucket. Anyone with a public bucket URL can access an object in that public bucket. If you are using a custom domain to expose the R2 bucket, you can manage authorization and access controls as you would for a Cloudflare zone. Public buckets only provide `GET`/`HEAD` on a known object path. Public bucket errors are rendered as HTML pages. There is no ability to do S3 operations such as `CreateBucket`, `ListBuckets`, or `DeleteBucket`.

Author: ```suggestion
[Public buckets](/r2/data-access/public-buckets/) are available on a regular HTTP endpoint. By default, there is no authorization or access controls associated with a public bucket. Anyone with a public bucket URL can access an object in that public bucket. If you are using a custom domain to expose the R2 bucket, you can manage authorization and access controls as you would for a Cloudflare zone. Public buckets only provide `GET`/`HEAD` on a known object path. Public bucket errors are rendered as HTML pages. There is no ability to do S3 operations such as `CreateBucket`, `ListBuckets`, or `DeleteBucket` with public buckets.
```

---

# Presigned URLs

Presigned URLs are an [S3 concept](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html) for sharing direct access to your bucket without revealing your token secret. A presigned URL authorizes anyone with the URL to perform an action to the S3 compatibility endpoint for an R2 bucket. By default, the S3 endpoint requires an `AUTHORIZATION` header signed by your token. Every presigned URL has S3 parameters and search parameters containing the signature information that would be present in an `AUTHORIZATION` header. The performable action is restricted to a specific resource, an [operation](/r2/data-access/s3-api/api/), and has an associated timeout.

There are three kinds of resources in R2: 

1. **Account**: For account-level operations (such as `CreateBucket`, `ListBuckets`, `DeleteBucket`) the identifier is the account ID.
2. **Bucket**: For bucket-level operations (such as `ListObjects`, `PutBucketCors`) the identifier is the account ID, and bucket name.
3. **Object**: For object-level operations (such as `GetObject`, `PutObject`, `CreateMultipartUpload`) the identifier is the account ID, bucket name, and object path.

> All parts of the identifier are part of the presigned URL. You cannot change the resource being accessed after the request is signed. For example, trying to change the bucket name to access the same object in a different bucket will return a `403` with an error code of `SignatureDoesNotMatch`.

Author: ```suggestion
All parts of the identifier are part of the presigned URL. 

You cannot change the resource being accessed after the request is signed. For example, trying to change the bucket name to access the same object in a different bucket will return a `403` with an error code of `SignatureDoesNotMatch`.
```

> Presigned URLs must have a defined expiry. You can set a timeout from one second to 7 days (604,800 seconds) into the future. The URL will contain the time when the URL was generated (`X-Amz-Date`) and the timeout (`X-Amz-Expires`) as search parameters. These search parameters are signed so tampering with them will result in `403` with an error code of `SignatureDoesNotMatch`.

Author: ```suggestion
Presigned URLs must have a defined expiry. You can set a timeout from one second to 7 days (604,800 seconds) into the future. The URL will contain the time when the URL was generated (`X-Amz-Date`) and the timeout (`X-Amz-Expires`) as search parameters. These search parameters are signed and tampering with them will result in `403` with an error code of `SignatureDoesNotMatch`.
```

In a Worker, you would change your upload to:

```ts
const existingObject = await env.DROP_BOX_BUCKET.put(
	url.toString().substring(1),
	request.body,
	{
		onlyIf: {
			// No objects will have been uploaded before September 28th, 2021 which
			// is the initial R2 announcement.
			uploadedBefore: new Date(1632844800000),
		},
	}
);
if (existingObject?.etag !== request.headers.get('etag')) {
	return new Response('attempt to overwrite object', { status: 400 });
}
```

Cloudflare Workers currently have some limitations that you may need to consider:

* You cannot upload more than 100 MiB (200 MiB for Business customers) to a Worker.
* Enterprise customers can upload 500 MiB by default and can ask their account team to raise this limit.
* Detecting [precondition failures](/r2/data-access/s3-api/extensions/#conditional-operations-in-putobject) is currently easier with presigned URLs as compared with R2 bindings.

Note that these limitations depends on R2's extension for conditional uploads. Amazon's S3 service does not offer such functionality at this time.

> ## Presigned URLs versus Public Buckets

Author: ```suggestion
## Presigned URLs versus public buckets
```
